contracts/src/v0.8/dev/keeper2_0/KeeperRegistrar2_0.sol

```diff
@@ -318,6 +318,40 @@ contract KeeperRegistrar2_0 is TypeAndVersionInterface, ConfirmedOwner, ERC677Re
     }
   }
 
+  /**
+   * @notice Allows external users to register upkeeps
+   * @notice registerUpkeep can only be called through transferAndCall on LINK contract
+   * @param name string of the upkeep to be registered
+   * @param upkeepContract address to perform upkeep on
+   * @param gasLimit amount of gas to provide the target contract when performing upkeep
+   * @param adminAddress address to cancel upkeep and withdraw remaining funds
+   * @param checkData data passed to the contract when checking for upkeep
+   * @param amount quantity of LINK upkeep is funded with (specified in Juels)
+   */
+  function registerUpkeep(
+    string memory name,
+    address upkeepContract,
+    uint32 gasLimit,
+    address adminAddress,
+    bytes calldata checkData,
+    uint96 amount
+  ) external returns (uint256) {
+    KeeperRegistryBaseInterface keeperRegistry = s_config.keeperRegistry;
+
+    uint256 upkeepId = keeperRegistry.registerUpkeep(upkeepContract, gasLimit, adminAddress, checkData);
+
+    bool success = LINK.transferFrom(msg.sender, owner(), amount);
+    if (!success) {
+      revert LinkTransferFailed(address(keeperRegistry));
+    }
+    
+    bytes32 hash = keccak256(abi.encode(upkeepContract, gasLimit, adminAddress, checkData));
+
+    emit RegistrationApproved(hash, name, upkeepId);
+
+    return upkeepId;
+  }
```

Comment:
can we reuse the register function for all registration logic? 

So it will be something like

if (amount < s_config.minLINKJuels) {
      revert InsufficientPayment();
    }
   bool success = LINK.transferFrom(msg.sender, owner(), amount);
     if (!success) {
       revert LinkTransferFailed(address(this));
     }

_register()

Reply:
I don't fully know why existing register function is external, i think it can be converted to internal and called from `onTokenTransfer` or `registerUpkeep` function

Reply:
The next commit addresses this change. There was a lot of common logic, so I did move that to a private function.